### ISSUE TITLE
(Java) Source RCON Sockets don't seem to be getting closed

### ISSUE THREAD

**Issue by comanoodle** [koraktor/steam-condenser#15]
This is related to my previous bug submission (the thread that clears passwords). After a while, I start getting this error whenever it attempts to run rcon exec().

My code is written such that each Server object (homemade object) has its own SourceServer object. Server object are created for every server when the application is started. They are held inside variables that remain in memory. When something attempts to issue a RCON command to that server, the rcon command checks if the SourceServer object was created, and if not, it creates it. If it was, it simply sends server.rconExec. All RCON commands to that server go through that Server's SourceServer object (in other words: only a single SourceServer object is ever created, and auth only happens once).

The SourceServer object (and auth) is not done inside the Server() function because I wanted the execute() function to be the one throwing the exceptions.

```
public String[] exec(String command) throws IOException, TimeoutException, SteamCondenserException, RCONNoAuthException {
    if (this.source == null) {
        InetAddress serverIp = InetAddress.getByName(server.getIP());
    source = new SourceServer(serverIp, server.getPort());
    source.rconAuth(server.getRconPassword());
    }


    String[] result = this.source.rconExec(command).split("\r\n|\r|\n");
    return result;
}
```

The only thing that repeats is the SteamCondenser exec(), so it's most likely that the exec sockets are left open somewhere, and those accumulate until no more sockets can be opened.

```
java.io.IOException: Too many open files
    at sun.nio.ch.IOUtil.initPipe(Native Method)
    at sun.nio.ch.EPollSelectorImpl.<init>(EPollSelectorImpl.java:67)
    at sun.nio.ch.EPollSelectorProvider.openSelector(EPollSelectorProvider.java:36)
    at java.nio.channels.Selector.open(Selector.java:227)
    at steamcondenser.steam.sockets.SteamSocket.receivePacket(SteamSocket.java:86)
    at steamcondenser.steam.sockets.RCONSocket.getReply(RCONSocket.java:48)
    at steamcondenser.steam.servers.SourceServer.rconExec(SourceServer.java:76)
    at coma.ntpugbot.ntRcon.exec(ntRcon.java:35)
    at coma.ntpugbot.Server.execute(Server.java:43)
    at coma.ntpugbot.Server.getRealPassword(Server.java:117)
    at coma.ntpugbot.Server.matchInProgress(Server.java:132)
    at coma.ntpugbot.Server.clearPassword(Server.java:61)
    at coma.ntpugbot.ClearPasswords.run(ClearPasswords.java:18)
```

**Comment by koraktor** [comment]
Sockets should not be the problem as the sockets are reused. The problem seems to be the Selector used to check for incoming data. As you might see from the backtrace, the exception is thrown when calling Selector.open().
I just forgot to call close() on the Selector in SteamSocket.receivePacket().

**Comment by comanoodle** [comment]
Well, I'll admit I didn't really know what was happening, so I simply made the first assumption that was logical to me :p

How (where) can I fix this? - or, are you going to be releasing a new version soon?

**Comment by koraktor** [comment]
This has been fixed in 818f94aeca81d7a0162cc8287ba15096f1da28d4.

If you need this now, you can apply this fix really easy by yourself. Otherwise wait for the new version, it will be released soon.

**Comment by comanoodle** [comment]
Great!

Thanks for all your hard work. SteamCondenser is awesome!
